### ISSUE TITLE
Replace `CUDA_VERSION` with `cudaRuntimeGetVersion()` for `if`/`IF` conditionals

### ISSUE THREAD

**Issue by kmaehashi** [cupy/cupy#5696]
This is needed to support CUDA Python as `CUDA_VERSION` in CUDA Python does not mean the current CUDA version.

**Comment by takagi** [comment]
`CUDA_VERSION` in Cython files?

**Comment by takagi** [comment]
```
$ find . -name '*.pyx' -or -name '*.px[id]' | xargs grep CUDA_VERSION
./cupy/cuda/memory.pyx:    if CUDA_VERSION < 11020:
./cupy/fft/_callback.pyx:                        '-E', 'CUDA_VERSION=' + _build_ver,
./cupy_backends/cuda/api/_driver_extern.pxi:    int CUDA_VERSION
./cupy_backends/cuda/api/_runtime_typedef.pxi:    IF CUDA_VERSION > 0:
./cupy_backends/cuda/api/_runtime_typedef.pxi:    IF CUDA_VERSION >= 11000:
./cupy_backends/cuda/api/_runtime_typedef.pxi:    ELIF CUDA_VERSION >= 10000:
./cupy_backends/cuda/api/_runtime_typedef.pxi:    ELIF CUDA_VERSION == 9020:
./cupy_backends/cuda/api/runtime.pyx:    IF CUDA_VERSION > 0 or HIP_VERSION > 0:
./cupy_backends/cuda/api/runtime.pyx:    IF CUDA_VERSION >= 9020:
./cupy_backends/cuda/api/runtime.pyx:    IF CUDA_VERSION >= 10000:
./cupy_backends/cuda/api/runtime.pyx:    IF CUDA_VERSION >= 11000:
./cupy_backends/cuda/api/runtime.pyx:    if CUDA_VERSION < 11020:
./cupy_backends/cuda/api/runtime.pyx:    if CUDA_VERSION < 11020:
./cupy_backends/cuda/api/runtime.pyx:    IF CUDA_VERSION > 0:
./cupy_backends/cuda/api/runtime.pyx:    if CUDA_VERSION < 11020:
./cupy_backends/cuda/api/runtime.pyx:    if CUDA_VERSION < 11020:
./cupy_backends/cuda/api/runtime.pyx:    if CUDA_VERSION < 11020:
./cupy_backends/cuda/api/runtime.pyx:    if CUDA_VERSION < 11020:
./cupy_backends/cuda/api/runtime.pyx:    if CUDA_VERSION < 11020:
./cupy_backends/cuda/api/runtime.pyx:    if CUDA_VERSION < 11020:
./cupy_backends/cuda/api/runtime.pyx:IF CUDA_VERSION >= 10000:
./cupy_backends/cuda/api/runtime.pyx:    if CUDA_VERSION < 10000:
./cupy_backends/cuda/api/driver.pyx:    # Note: CUDA_VERSION is defined either in CUDA Python or _driver_extern.pxi
./cupy_backends/cuda/api/driver.pyx:    if CUDA_VERSION > 0:
./cupy_backends/cuda/api/driver.pyx:        return CUDA_VERSION
./cupy_backends/cuda/api/driver.pyx:    if CUDA_VERSION < 9000:
./cupy_backends/cuda/libs/nvrtc.pyx:    if CUDA_VERSION < 11020 or runtime._is_hip_environment:
./cupy_backends/cuda/libs/nvrtc.pyx:    if CUDA_VERSION < 11010 or runtime._is_hip_environment:
```

**Comment by kmaehashi** [comment]
Yes, but `_runtime_typedef.pxi` can be kept as-is as `IF CUDA_VERSION` condition cannot be removed and it is not used by CUDA Python.

**Comment by takagi** [comment]
Seems some codes depend on Cython's dead code elimination:

https://github.com/cupy/cupy/blob/f81c85f1239d1b3da3fb3d7be3dd701b988526da/cupy_backends/cuda/api/driver.pyx#L278-L286

https://github.com/cupy/cupy/blob/f81c85f1239d1b3da3fb3d7be3dd701b988526da/cupy_backends/cuda/api/runtime.pyx#L771-L777

As only a few modules, `driver`, `runtme` and `nvrtc`, import CUDA Python, giving an alias to `CUDA_VERSION` in a C header and importing it may be a handy way.

**Comment by kmaehashi** [comment]
You can remove these conditionals as we no longer support CUDA 9.x 😄 

Besides that, in general, build-time conditionals (`IF`) must be replaced with `if`, as the CuPy built with CUDA Python will not depend on a specific CUDA version.

> As only a few modules, driver, runtme and nvrtc, import CUDA Python, giving an alias to CUDA_VERSION in a C header and importing it may be a handy way.

We are going to use CUDA Python in other modules e.g. cublas once CUDA Python support is available

**Comment by takagi** [comment]
`get_build_version` wants `CUDA_VERSION` at the build time. How about this?

https://github.com/cupy/cupy/blob/f81c85f1239d1b3da3fb3d7be3dd701b988526da/cupy_backends/cuda/api/driver.pyx#L78-L79

**Comment by kmaehashi** [comment]
I think it's ok to leave it as-is. `show_config()` uses it like this:
https://github.com/cupy/cupy/blob/f81c85f1239d1b3da3fb3d7be3dd701b988526da/cupyx/_runtime.py#L152-L159

Other use of `get_build_version()` may need to be rewritten.

**Comment by takagi** [comment]
Okay, we still have the build-time CUDA version so we need to see and bake `CUDA_VERSION` in `cuda.h` there instad of CUDA Python's somehow, right? Also, I'll check where `get_build_version()` is used.

**Comment by kmaehashi** [comment]
Yes. Currently only `{driver,runtime,nvrtc}.pyx` uses CUDA Python, whereas other modules still directly link to CUDA Toolkit libraries. So we still need to burn-in build-time `CUDA_VERSION` from the `cuda.h` (not CUDA Python). However CUDA Python also exports `CUDA_VERSION`, now we have to do this bunr-in somewhere else than `driver.pyx`.

**Comment by kmaehashi** [comment]
My idea is to do that in every module using `driver.get_build_version()`. For example, burn-in `CUDA_VERSION` in `cupy_backends.cuda.libs.cusolver`, then use it in 
https://github.com/cupy/cupy/blob/f81c85f1239d1b3da3fb3d7be3dd701b988526da/cupy/cusolver.pyx#L109

**Comment by takagi** [comment]
Let me clarify that `driver.get_build_version()` returns CUDA Python's `CUDA_VERSION` if it is left as it-is (when CUDA Python is enabled). So I think we need to change `driver.get_build_version()` somehow to see `cuda.h`'s `CUDA_VERSION`.

Also, I think making each module have the burn-in `CUDA_VERSION` is reasonable, though I'm not sure `driver.get_build_version()` is expanded to `CUDA_VERSION` at build-time and it may be left as a runtime call. How about directly declaring extern `CUDA_VERSION` from `cupy_backend.h` to make it burned-in at each module?

**Comment by takagi** [comment]
Ah, let me make myself understand the case further...

**Comment by kmaehashi** [comment]
> Let me clarify that `driver.get_build_version()` returns CUDA Python's `CUDA_VERSION` if it is left as it-is (when CUDA Python is enabled). So I think we need to change `driver.get_build_version()` somehow to see `cuda.h`'s `CUDA_VERSION`.

Also, I agree that making each module have the burn-in `CUDA_VERSION` is reasonable, though I'm not sure `driver.get_build_version()` is expanded to `CUDA_VERSION` at build-time and it may be left as a runtime call. How about directly declaring extern `CUDA_VERSION` from `cupy_backend.h` to make it burned-in at each module?

**Comment by kmaehashi** [comment]
> Let me clarify that `driver.get_build_version()` returns CUDA Python's `CUDA_VERSION` if it is left as it-is (when CUDA Python is enabled).

Yes.

> So I think we need to change `driver.get_build_version()` somehow to see `cuda.h`'s `CUDA_VERSION`.

I think it's better to expose `CUDA_VERSION` as `cupy_backends.cuda.libs.cusolver._get_cuda_build_version()` etc.

**Comment by takagi** [comment]
Phew, now I got your point about how `driver.get_build_version` should work. I misunderstood that.

**Comment by takagi** [comment]
In an offline talk, we found that Cython looks up and finds `cuda.h`'s `CUDA_VERSION` first even when `cuda.ccuda.*` is cimported.

**Comment by kmaehashi** [comment]
Let's prefix all `compile_time_env`s with `CUPY_` to avoid confusion and conflict like this 😅 
https://github.com/cupy/cupy/blob/master/cupy_setup_build.py#L947-L957
